### PR TITLE
Use Payload API for Events

### LINKED ISSUE #318: Event Reworking
The specs for the current system are not quite correct. With issues across many areas of the section, I think it is best that we start this section again and have another go at it. I will put all the requirements on this issue for now, but will split into separate issues when we get round to working on it.

The events system should be able to handle BYBA's competitive events as well as concerts etc put on by member bands. 

Events should have several levels of visibility: 
1. Draft - this is visible to the user that created the event only
2. Internal - this would allow the event to be viewed on `OrgAdmin` but not released publicly (e.g. allowing bands to register interest/entries before dates go live)
3. Published - this is when the event can be regarded as "public"
4. Complete - the event is viewable publicly AND the scores have been released

There are certain criteria needed to allow events to move to the next visibility level:
- Draft -> Internal needs an event name, date and address
- Internal -> Published needs at least 2 entered bands (this check can be skipped for non-competitive events)
- Published -> Complete needs scores for each band (this check can be skipped for non-competitive events)

Competitive events need to allow for band registration by division. There should be a schedule which links to the entered bands (currently it is all free text). Bands should also be able to enter in exhibition, allowing for them to be inserted anywhere in the schedule.

The draw should be separated from the schedule. BYBA is responsible for carrying out the draw but the show organiser is responsible for scheduling bands.

There is no longer a concept of "Members-only" shows, so we do not need this field.

Judges can be assigned to the whole show, or to just one specific division

### LINKED ISSUE #336: Migrate Events section over to Payload
The events section should use `payload` as the backing source for its data, instead of our local database

### PR DESCRIPTION
**Pull request by rugulous**
Close #336 by removing all the existing event pages and pointing them to look straight at Payload

There are still some DB tables that need dropping, but no harm in leaving them in for now.

Also begins an API wrapper that we can *hopefully* use elsewhere :)

### CODE REVIEW

**Comment by rugulous** [comment]
‼️ don't approve this yet!‼️

We need to add Dev./admin.byba.online to the allowed domains!

**Comment by rugulous** [comment]
This also closes #318 by adding the different statuses and validation

**Comment by matt-fidd** [comment]
Where do I stand on reviewing this, can I have a look at it if I add the dev url to the CORS allowed list?

As an aside, do wildcards work in CORS, could I add *.byba.online to the allowed list...?

**Comment by rugulous** [comment]
@matt-fidd sorry just seen your comment!

By all means review it, it just won't work if we merge it anywhere lmao (the dev site is currently broken)

[From what I can see, it's all or nothing ](https://stackoverflow.com/a/14006226/9034824)- either `*` to allow everything, `null` to allow nothing, or *the exact protocol + domain + port* (what a ballache!). But that's the `CORS` spec itself, not necessarily payload's approach!

Assuming payload/CORS have thee same approach, we could probably do with having a chat about the pros/cons of enabling it for everything. **IN THEORY** there isn't a CORS issue with people making requests from any domain (e.g. if bands wanted to put a list of events on their own site or display their results), but there may be issues with CSRF if we allow any site to mess with data using a logged in user's cookie. Payload does fortunately distinguish the two so it may even be that we allow anyone to make a CORS request, but lock down accepting the cookie to just byba domains?

**Comment by rugulous** [comment]
@matt-fidd what you think?

**Comment by matt-fidd** [comment]
Looking good! Would you like me to add it to CORS and we can get it merged in?

File: views/event/view.hbs [hunk 1]
```diff
@@ -1,178 +1,184 @@
...
+			},
+			validate: function () {
+				this.errors = [];
+
+				if(!this.canEdit){
+					this.errors.push("You do not have permission to edit this event");
+					return false;
+				}
+
+				if (this.event.status == "draft") {
```
Comment (matt-fidd) [inline views/event/view.hbs:153]: `===`?

**Review by matt-fidd — APPROVED** [review]

**Comment by matt-fidd** [comment]
Also looking to fix the dates in the event to proper actual date types once I can get my head around it - I hate dates...

**Comment by rugulous** [comment]
Go for it 🫡

I'm not going to publish it live just yet though, until we've had a chance to have a think about whether we want to update the main BYBA site at payload too!

**Comment by rugulous** [comment]
@matt-fidd I've tweaked this a little to look at the event changes you'd made

Biggest change is that, I quite like the way we'd restricted events to starting and finishing on the same day. So I've updated the `byba-sdk` (API wrapper) to take the response and reformat it in the way the existing code was expecting, such that we still have a `date` and a `startTime`/`endTime` even though the backend is now storing these in a much nicer way

Happy to take feedback on this approach, but I think it gives us more flexibility if we were to decide later down the line that we wanted to support e.g. weekend-long events instead of just those spanning 1 day

File: views/layouts/main.hbs [hunk 2]
```diff
@@ -1,175 +1,175 @@
...
-	<main>
-		<nav class="navbar d-lg-none navbar-dark bg-default">
-			<div class="container-fluid">
-				<a class="navbar-brand" href="#">{{getConfigValue 'config/site' 'title' 'Org' }}</a>
-				<button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#sidebar" aria-controls="sidebar" aria-expanded="false" aria-label="Toggle navigation">
-					<span class="navbar-toggler-icon"></span>
-				</button>
-			</div>
-		</nav>
+	<body>
```
Comment (matt-fidd) [inline views/layouts/main.hbs:11]: I'm not a fan of the indentation level here personally, I tend to throw HTML and HEAD/BODY at the same level just to pull everything into the margin. But that's a personal choice - feel free to completely ignore me!

File: views/event/index.hbs [hunk 3]
```diff
@@ -62,18 +62,23 @@
...
+
+				byba.events.list().then(response => {
+					this.events = response.json.docs;
+					this.loading = false;
+				});
+			},
+
+			formatDate(dateStr){
+				const date = new Date(dateStr);
+				return date.toLocaleDateString(navigator.language);
```
Comment (matt-fidd) [inline views/event/index.hbs]: Nice!

**Review by matt-fidd — APPROVED** [review]

File: views/layouts/main.hbs (hunk 2 above)
Comment (rugulous) [reply views/layouts/main.hbs:11]: If I'm honest, I hadn't actually changed anything here but think I got VS to format for me... will be a ballache to have to constantly fight against auto-tools so I'm going to leave this for now!

**Comment by rugulous** [comment]
@matt-fidd I've made changes since you last reviewed, but if you're happy then I think I'm going to merge this branch and create a new one for band registration etc

**Pull request merged**